# Loading Libraries and Data

In [1]:
import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense

In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784  # 28*28
output_dim = nb_classes = 10
nb_epoch = 20

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

11501568/11490434 [==============================] - 0s 0us/step


In [3]:
Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

# Task 1: In this task, you'll implement several ANN models with different batch sizes.

## 1. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use 8 as the mini-batch size.

In [4]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=8, epochs=20, verbose=1)

Epoch 1/20
7500/7500 [==============================] - 16s 2ms/step - loss: 0.3644 - accuracy: 0.8946
Epoch 2/20
7500/7500 [==============================] - 15s 2ms/step - loss: 0.1718 - accuracy: 0.9500
Epoch 3/20
7500/7500 [==============================] - 16s 2ms/step - loss: 0.1241 - accuracy: 0.9645
Epoch 4/20
7500/7500 [==============================] - 16s 2ms/step - loss: 0.0970 - accuracy: 0.9715
Epoch 5/20
7500/7500 [==============================] - 16s 2ms/step - loss: 0.0795 - accuracy: 0.9766
Epoch 6/20
7500/7500 [==============================] - 16s 2ms/step - loss: 0.0666 - accuracy: 0.9799
Epoch 7/20
7500/7500 [==============================] - 16s 2ms/step - loss: 0.0566 - accuracy: 0.9834
Epoch 8/20
7500/7500 [==============================] - 15s 2ms/step - loss: 0.0482 - accuracy: 0.9862
Epoch 9/20
7500/7500 [==============================] - 16s 2ms/step - loss: 0.0425 - accuracy: 0.9869
Epoch 10/20
7500/7500 [==============================] - 16s 2ms/step - l

In [5]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.07671058923006058
Test accuracy: 0.978600025177002


## 2. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use 128 as the mini-batch size.

In [6]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 4ms/step - loss: 1.3210 - accuracy: 0.6501
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.5310 - accuracy: 0.8640
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3987 - accuracy: 0.8909
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3457 - accuracy: 0.9041
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3145 - accuracy: 0.9121
Epoch 6/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2928 - accuracy: 0.9175
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2758 - accuracy: 0.9221
Epoch 8/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2616 - accuracy: 0.9263
Epoch 9/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2495 - accuracy: 0.9302
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2388 - accuracy: 0.9330

In [7]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.17072197794914246
Test accuracy: 0.9490000009536743


## 3. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the full sample as the batch size.

In [8]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=X_train.shape[0], epochs=20, verbose=1)

Epoch 1/20
1/1 [==============================] - 1s 1s/step - loss: 2.2930 - accuracy: 0.1451
Epoch 2/20
1/1 [==============================] - 1s 633ms/step - loss: 2.2884 - accuracy: 0.1499
Epoch 3/20
1/1 [==============================] - 1s 618ms/step - loss: 2.2839 - accuracy: 0.1543
Epoch 4/20
1/1 [==============================] - 1s 640ms/step - loss: 2.2794 - accuracy: 0.1595
Epoch 5/20
1/1 [==============================] - 1s 621ms/step - loss: 2.2751 - accuracy: 0.1649
Epoch 6/20
1/1 [==============================] - 1s 605ms/step - loss: 2.2708 - accuracy: 0.1702
Epoch 7/20
1/1 [==============================] - 1s 613ms/step - loss: 2.2665 - accuracy: 0.1748
Epoch 8/20
1/1 [==============================] - 1s 614ms/step - loss: 2.2623 - accuracy: 0.1807
Epoch 9/20
1/1 [==============================] - 1s 614ms/step - loss: 2.2582 - accuracy: 0.1861
Epoch 10/20
1/1 [==============================] - 1s 623ms/step - loss: 2.2541 - accuracy: 0.1913
Epoch 11/20
1/1 [=====

In [9]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 2.2111291885375977
Test accuracy: 0.2515999972820282


## 4. Compare the results of each model. Which batch size performed best?

128, 64, and 10 neurons in the layers, mini-batch size 8
- Train Score, Accuracy
 - 0.0080, 0.9987
- Test Score, Accuracy
 - 0.0767, 0.9786

128, 64, and 10 neurons in the layers, mini-batch size 128
- Train Score, Accuracy
 - 0.1715, 0.9512
- Test Score, Accuracy
 - 0.1707, 0.9490

128, 64, and 10 neurons in the layers, full size batch
- Train Score, Accuracy
 - 2.2150, 0.2458
- Test Score, Accuracy
 - 2.2111, 0.2516

The mini-batch size 8 had the best results in both the score and accuracy, but there is a significant difference in the differences in train and test score showing a potential for overfitting. Therefore, the model with a mini-batch size of 128 is the best model for this data.


# Task 2: In this task, you'll implement several ANN models with different learning rates for the stochastic gradient descent. In all of the models below, use 128 as your mini-batch size.

In [10]:
from tensorflow.keras import optimizers
sgd_001 = optimizers.SGD(lr=0.01)
sgd_100 = optimizers.SGD(lr=100)
sgd_00000001 = optimizers.SGD(lr=0.0000001)

## 1. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use 0.01 as the learning rate.

In [11]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer=sgd_001, loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 2s 5ms/step - loss: 1.2303 - accuracy: 0.7079
Epoch 2/20
469/469 [==============================] - 2s 4ms/step - loss: 0.4979 - accuracy: 0.8716
Epoch 3/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3904 - accuracy: 0.8928
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3452 - accuracy: 0.9038
Epoch 5/20
469/469 [==============================] - 2s 4ms/step - loss: 0.3173 - accuracy: 0.9106
Epoch 6/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2971 - accuracy: 0.9153
Epoch 7/20
469/469 [==============================] - 2s 4ms/step - loss: 0.2807 - accuracy: 0.9202
Epoch 8/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2667 - accuracy: 0.9247
Epoch 9/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2546 - accuracy: 0.9278
Epoch 10/20
469/469 [==============================] - 2s 5ms/step - loss: 0.2436 - accuracy: 0.9308

In [12]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.16960208117961884
Test accuracy: 0.9495999813079834


## 2. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use 100 as the learning rate.

In [13]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer=sgd_100, loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 3s 5ms/step - loss: 141194418999552638976.0000 - accuracy: 0.1049
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 40.0812 - accuracy: 0.0999
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 40.6658 - accuracy: 0.0989
Epoch 4/20
469/469 [==============================] - 2s 5ms/step - loss: 39.6073 - accuracy: 0.1025
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 40.8037 - accuracy: 0.0997
Epoch 6/20
469/469 [==============================] - 2s 5ms/step - loss: 39.5059 - accuracy: 0.1017
Epoch 7/20
469/469 [==============================] - 2s 5ms/step - loss: 40.2279 - accuracy: 0.1014
Epoch 8/20
469/469 [==============================] - 2s 5ms/step - loss: 40.2535 - accuracy: 0.1000
Epoch 9/20
469/469 [==============================] - 2s 5ms/step - loss: 40.7341 - accuracy: 0.0991
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - los

In [14]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 925.3594970703125
Test accuracy: 0.08919999748468399


## 3. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use 0.0000001 as the learning rate.

In [15]:
model = Sequential()
# our first dense layer
model.add(Dense(128, input_shape=(784,), activation="relu"))
# our second dense layer
model.add(Dense(64, activation="relu"))
# last layer is the output layer.
model.add(Dense(10, activation="softmax"))

model.compile(optimizer=sgd_00000001, loss='categorical_crossentropy',
              metrics=['accuracy'])

# setting verbose=1 prints out some results after each epoch
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1)

Epoch 1/20
469/469 [==============================] - 3s 5ms/step - loss: 2.3497 - accuracy: 0.0795
Epoch 2/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3497 - accuracy: 0.0795
Epoch 3/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3497 - accuracy: 0.0795
Epoch 4/20
469/469 [==============================] - 2s 4ms/step - loss: 2.3496 - accuracy: 0.0795
Epoch 5/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3496 - accuracy: 0.0795
Epoch 6/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3496 - accuracy: 0.0795
Epoch 7/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3496 - accuracy: 0.0795
Epoch 8/20
469/469 [==============================] - 2s 4ms/step - loss: 2.3496 - accuracy: 0.0795
Epoch 9/20
469/469 [==============================] - 2s 5ms/step - loss: 2.3496 - accuracy: 0.0795
Epoch 10/20
469/469 [==============================] - 2s 4ms/step - loss: 2.3496 - accuracy: 0.0795

In [16]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 2.3481955528259277
Test accuracy: 0.08420000225305557


## 4. Compare the results of each model. Which learning rate performed best?

128, 64, and 10 neurons, mini-batch size 128, learning rate 0.01
- Train Score, Accuracy
 - 0.1711, 0.9510
- Test Score, Accuracy
 - 0.1696, 0.9496

128, 64, and 10 neurons, mini-batch size 128, learning rate 100
- Train Score, Accuracy
 - 39.7179, 0.1018
- Test Score, Accuracy
 - 925.3595, 0.0892

128, 64, and 10 neurons, mini-batch size 128, learning rate 0.0000001
- Train Score, Accuracy
 - 2.3495, 0.0795
- Test Score, Accuracy
 - 2.3482, 0.0842

The learning rate 0.01 had the best results and converged. The learning rate 100 diverged because the value was too large. The learning rate 0.0000001 model improved very slowly causing minimal improvements in 20 epochs.